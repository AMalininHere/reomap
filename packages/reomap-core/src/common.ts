import { tile2lat, tile2lng, lng2tile, lat2tile } from './utils/geo-fns';

export class LatLng {
  constructor(
    public readonly lat: number,
    public readonly lng: number,
  ) { }

  equals(other: LatLng) {
    return (
      this.lat === other.lat &&
      this.lng === other.lng
    );
  }
}

export class Point {
  constructor(
    public readonly x: number,
    public readonly y: number,
  ) { }
}

export const TILE_SIZE = 256;

export const absMinLatLng = new LatLng(
  tile2lat(Math.pow(2, 10), 10),
  tile2lng(0, 10)
);

export const absMaxLatLng = new LatLng(
  tile2lat(0, 10),
  tile2lng(Math.pow(2, 10), 10)
);

export function pixelToLatLng(width: number, height: number, zoom: number, center: LatLng, source: Point) {
  const pointDiffX = (source.x - width / 2) / TILE_SIZE;
  const pointDiffY = (source.y - height / 2) / TILE_SIZE;

  const tileX = lng2tile(center.lng, zoom) + pointDiffX;
  const tileY = lat2tile(center.lat, zoom) + pointDiffY;

  return new LatLng(
    Math.max(absMinLatLng.lat, Math.min(absMaxLatLng.lat, tile2lat(tileY, zoom))),
    Math.max(absMinLatLng.lng, Math.min(absMaxLatLng.lng, tile2lng(tileX, zoom)))
  );
}

export function latLngToPixel(width: number, height: number, zoom: number, center: LatLng, source: LatLng) {
  const tileCenterX = lng2tile(center.lng, zoom);
  const tileCenterY = lat2tile(center.lat, zoom);

  const tileX = lng2tile(source.lng, zoom);
  const tileY = lat2tile(source.lat, zoom);

  return new Point(
    (tileX - tileCenterX) * TILE_SIZE + width / 2,
    (tileY - tileCenterY) * TILE_SIZE + height / 2
  );
}
