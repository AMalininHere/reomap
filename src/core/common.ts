import { tile2lat, tile2lng, lng2tile, lat2tile } from './utils/geo-fns';
import { LatLng, Point } from './models';

export const absMinLatLng = new LatLng(
  tile2lat(Math.pow(2, 10), 10),
  tile2lng(0, 10)
);

export const absMaxLatLng = new LatLng(
  tile2lat(0, 10),
  tile2lng(Math.pow(2, 10), 10)
);

export function pixelToLatLng(width: number, height: number, zoom: number, center: LatLng, source: Point) {
  const pointDiffX = (source.x - width / 2) / 256.0;
  const pointDiffY = (source.y - height / 2) / 256.0;

  const tileX = lng2tile(center.lng, zoom) + pointDiffX;
  const tileY = lat2tile(center.lat, zoom) + pointDiffY;

  return new LatLng(
    Math.max(absMinLatLng.lat, Math.min(absMaxLatLng.lat, tile2lat(tileY, zoom))),
    Math.max(absMinLatLng.lng, Math.min(absMaxLatLng.lng, tile2lng(tileX, zoom)))
  );
}

export function latLngToPixel(width: number, height: number, zoom: number, center: LatLng, source: LatLng) {
  const tileCenterX = lng2tile(center.lng, zoom);
  const tileCenterY = lat2tile(center.lat, zoom);

  const tileX = lng2tile(source.lng, zoom);
  const tileY = lat2tile(source.lat, zoom);

  return new Point(
    (tileX - tileCenterX) * 256.0 + width / 2,
    (tileY - tileCenterY) * 256.0 + height / 2
  );
};
