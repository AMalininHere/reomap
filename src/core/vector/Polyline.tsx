import React from 'react';
import { LatLng, Point } from '../models';

function makeSvgPath(points: Point[]) {
  return points
    .map((p, idx) => `${idx === 0 ? 'M' : 'L'} ${p.x} ${p.y}`)
    .join(' ');
}

export interface Props {
  positions: LatLng[];
  latLngToPixel: (ll: LatLng) => Point;
}

function Polyline(props: Props) {
  const {
    positions,
    latLngToPixel,
  } = props;

  const pathString = makeSvgPath(positions.map(latLngToPixel));

  return (
    <path fill="none" stroke="#555555" strokeWidth={2} d={pathString} />
  );
}

export default React.memo(Polyline);