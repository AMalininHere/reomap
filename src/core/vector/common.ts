import * as G from 'geojson';
import { LatLng, Point } from '../models';

export interface ElementProps<T extends G.Geometry> {
  geoElement: T;
  latLngToPixel: (ll: LatLng) => Point;
}