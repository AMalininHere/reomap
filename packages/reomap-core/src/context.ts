import React, { useContext } from 'react';
import { LatLng, Point, latLngToPixel, pixelToLatLng } from './common';

export class ContextState {
  constructor(
    public readonly center: LatLng,
    public readonly zoom: number,
    public readonly width: number,
    public readonly height: number,
  ) { }

  public readonly latLngToPixel = (source: LatLng) =>
    latLngToPixel(this.width, this.height, this.zoom, this.center, source);

  public readonly pixelToLatLng = (source: Point) =>
    pixelToLatLng(this.width, this.height, this.zoom, this.center, source);
}

const ctx = React.createContext<ContextState>(null!);

export const MapProvider = ctx.Provider;

export function useMapContext() {
  const data = useContext(ctx);

  return data;
}
