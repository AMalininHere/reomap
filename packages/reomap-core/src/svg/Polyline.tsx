import React, { Ref } from 'react';
import { useMapContext } from '../context';
import { LatLng, Point } from '../common';

function makeSvgPath(points: Point[]) {
  if (!points.length) {
    return '';
  }

  let p = points[0];
  let x = Math.round(p.x);
  let y = Math.round(p.y);

  const resultItems = [`M${x} ${y}`];

  for (let i = 1; i < points.length; i++) {
    p = points[i];

    const nextX = Math.round(p.x);
    const nextY = Math.round(p.y);

    if (nextX === x && nextY === y) {
      continue;
    }

    x = nextX;
    y = nextY;
    resultItems.push(`L${x} ${y}`);
  }

  return resultItems.join(' ');

}

interface Props {
  positions: LatLng[];
}

type PathProps = Omit<React.SVGProps<SVGPathElement>, 'fill' | 'd'>;

function Polyline(props: Props & PathProps, ref: Ref<SVGPathElement>) {
  const {
    positions,
    ...pathProps
  } = props;
  const mapContext = useMapContext();

  const pathString = makeSvgPath(positions.map(mapContext.latLngToPixel));

  const pathPropsWitdhDefaults: PathProps = {
    stroke: '#555555',
    strokeWidth: 2,
    ...pathProps
  };

  return (
    <path ref={ref} fill="none" d={pathString} {...pathPropsWitdhDefaults} />
  );
}

export default React.memo(React.forwardRef(Polyline));
