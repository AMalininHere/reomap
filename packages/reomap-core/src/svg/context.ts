import { createContext, useContext } from 'react';
import { ContextState } from '../context';

const ctx = createContext<ContextState>(null!);
ctx.displayName = 'SvgLayerContext';

export const SvgLayerProvider = ctx.Provider;

export function useSvgLayerContext() {
  const data = useContext(ctx);

  return data;
}
