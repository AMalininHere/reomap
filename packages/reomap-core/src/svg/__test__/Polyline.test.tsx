import { render } from '@testing-library/react';
import Polyline from '../Polyline';
import { SvgLayerProvider } from '../context';
import { createContextState } from '../../context';
import { latLng } from '../../common';

describe('Polyline', () => {
  it('should render', () => {
    const contextState = createContextState(latLng(55.417, 85.276), 6, 800, 600);

    const geojsonLinePath: [number, number][] = [
      [
        82.935791015625,
        55.02802211299252
      ],
      [
        84.8583984375,
        55.78892895389262
      ],
      [
        86.099853515625,
        55.36662484928637
      ],
      [
        86.209716796875,
        54.67383096593114
      ],
      [
        86.737060546875,
        53.93021986394
      ],
      [
        87.154541015625,
        53.72271667491848
      ]
    ];

    const { container } = render(
      <svg>
        <SvgLayerProvider value={contextState}>
          <Polyline positions={geojsonLinePath.map(([lng, lat]) => latLng(lat, lng))}/>
        </SvgLayerProvider>
      </svg>
    );

    expect(container).toMatchSnapshot();
  });
});
