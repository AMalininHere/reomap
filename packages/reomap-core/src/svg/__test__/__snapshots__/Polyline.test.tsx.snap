// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Polyline should render 1`] = `
<div>
  <svg>
    <path
      d="M293 331 L381 270 L437 304 L442 359 L466 417 L485 433"
      fill="none"
      stroke="#555555"
      stroke-width="2"
    />
  </svg>
</div>
`;
